src/odemis/driver/actuator.py

```diff
@@ -940,6 +940,203 @@ def reference(self, axes):
+
+        logging.debug("Subscribing to position of child %s", child.name)
+        child.position.subscribe(self._update_child_position, init=True)
+
+        if model.hasVA(child, "referenced") and axis_name in child.referenced.value:
+            self._referenced[axis] = child.referenced.value[axis_name]
+            self.referenced = model.VigilantAttribute(self._referenced.copy(), readonly=True)
+            child.referenced.subscribe(self._update_child_ref)
+
+        # If the axis can be referenced => do it now (and move to a known position)
+        # In case of cyclic move always reference
+        if not self._referenced.get(axis, True):
+            # The initialisation will not fail if the referencing fails
+            f = self.reference({axis})
+            f.add_done_callback(self._on_referenced)
+        else:
+            self.moveAbs({self._axis: self._child.position.value[self._caxis]}).result()
+
+    def _on_referenced(self, future):
+        try:
+            future.result()
+        except Exception as e:
+            self._child.stop({self._caxis})  # prevent any move queued
+            self.state._set_value(e, force_write=True)
+            logging.exception(e)
+
+    def _update_child_position(self, value):
+        p = value[self._caxis]
+        self._position[self._axis] = p - self._offset
+        self._updatePosition()
+
+    def _update_child_ref(self, value):
+        self._referenced[self._axis] = value[self._caxis] - self._offset
+        self._updateReferenced()
+
+    def _updatePosition(self):
+        """
+        update the position VA
+        """
+        # if it is an unsupported position report the nearest supported one
+        pos = self._position[self._axis]
+        self.position._set_value({self._axis: pos}, force_write=True)
+
+    def _updateReferenced(self):
+        """
+        update the referenced VA
+        """
+        # .referenced is copied to detect changes to it on next update
+        self.referenced._set_value(self._referenced.copy(), force_write=True)
+
+    @isasync
+    def moveRel(self, shift):
+        """
+        Move the rotation actuator by a defined shift.
+        shift dict(string-> float): name of the axis and shift
+        """
+        if not shift:
+            return model.InstantaneousFuture()
+        self._checkMoveRel(shift)
+        f = self._executor.submit(self._doMoveRel, shift)
+
+        return f
+
+    def _doMoveRel(self, shift):
+        """
+        shift dict(string-> float): name of the axis and shift
+        """
+        cur_pos = self._child.position.value[self._caxis] - self._offset
+        pos = cur_pos + shift[self._axis]
+        self._doMoveAbs({self._axis: pos})
+
+    @isasync
+    def moveAbs(self, pos):
+        """
+        Move the actuator to the defined position in m for each axis given.
+        pos dict(string-> float): name of the axis and position in m
+        """
+        self._move_num_total += 1
```

Comment:
This needs to be done in the `_doMoveAbs()`. `moveAbs` should always stay asynchronous.